src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +115,506 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                }
+
+                return null;
+            }
+
+            /// <summary>
+            /// Returns method: void Dispose(bool)
+            /// </summary>
+            private static IMethodSymbol FindDisposeBoolMethod(INamedTypeSymbol type)
+            {
+                return type.GetMembers(DisposeMethodName).OfType<IMethodSymbol>().FirstOrDefault(IsDisposeBoolMethod);
+            }
+        }
+
+        /// <summary>
+        /// Validates implementation of Dispose method. The method must call Dispose(true) and then GC.SuppressFinalize(this).
+        /// </summary>
+        private struct DisposeImplementationValidator
+        {
+            // this type will be created per compilation 
+            // this is actually a bug - https://github.com/dotnet/roslyn-analyzers/issues/845
+#pragma warning disable RS1008
+            private readonly IMethodSymbol suppressFinalizeMethod;
+            private readonly INamedTypeSymbol type;
+#pragma warning restore RS1008
+            private bool callsDisposeBool;
+            private bool callsSuppressFinalize;
+
+            public DisposeImplementationValidator(IMethodSymbol suppressFinalizeMethod, INamedTypeSymbol type)
+            {
+                callsDisposeBool = false;
+                callsSuppressFinalize = false;
+                this.suppressFinalizeMethod = suppressFinalizeMethod;
+                this.type = type;
+            }
+
+            public bool Validate(ImmutableArray<IOperation> operations)
+            {
+                callsDisposeBool = false;
+                callsSuppressFinalize = false;
+
+                if (ValidateOperations(operations))
+                {
+                    return callsDisposeBool && callsSuppressFinalize;
+                }
+
+                return false;
+            }
+
+            private bool ValidateOperations(ImmutableArray<IOperation> operations)
+            {
+                foreach (var operation in operations)
+                {
+                    if (!ValidateOperation(operation))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private bool ValidateOperation(IOperation operation)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.EmptyStatement:
+                    case OperationKind.LabelStatement:
+                        return true;
+                    case OperationKind.BlockStatement:
+                        var blockStatement = (IBlockStatement)operation;
+                        return ValidateOperations(blockStatement.Statements);
+                    case OperationKind.ExpressionStatement:
+                        var expressionStatement = (IExpressionStatement)operation;
+                        return ValidateExpression(expressionStatement);
+                    default:
```

Comment:
cc @JohnHamby: This would be an example of an analyzer depending on the statement structure imposed by VB/C#. I'm trying to think about how this might be made less fragile (possibly using OperationWalker instead of manual switch), but it seems very to implement a "do X and nothing else" check and not have this language-affinity.

Don't block on this as PR feedback, but it's something for us to consider medium term.